builder/cse.cc

```diff
@@ -271,6 +271,29 @@ class graph_node_mutator : public node_mutator {
 
 // ----------------------
 
+// Only pure exprs should be considered for CSE.
+bool is_pure(const expr& e) {
+  if (const call* c = e.as<call>()) {
+    switch (c->intrinsic) {
+      // buffer accessors and semaphore helpers are never pure.
+    case intrinsic::buffer_rank:
+    case intrinsic::buffer_elem_size:
+    case intrinsic::buffer_size_bytes:
+    case intrinsic::buffer_min:
+    case intrinsic::buffer_max:
+    case intrinsic::buffer_stride:
+    case intrinsic::buffer_fold_factor:
+    case intrinsic::buffer_at:
+    case intrinsic::semaphore_init:
+    case intrinsic::semaphore_signal:
+    case intrinsic::semaphore_wait: return false;
+    default: return true;
```

Comment:
`free`, `trace_begin`, and `trace_end` aren't pure either

Reply:
Cool, will add

Reply:
It would be better to flip this around and return true when the intrinsic is pure and false otherwise. That way, when someone adds an intrinsic, and they forget to mark it as pure, the failure is lack of CSE, rather than likely very hard to find bugs.